suave/contributing/writing-a-precompile.md

```diff
@@ -0,0 +1,119 @@
+
+# Writing a precompile
+
+SUAVE uses custom precompiles to extend the EVM with specific MEV functions.
+
+It uses a specification in `yaml` to describe each precompile with its inputs and outputs. From this specification, two bindings are auto-generated, one in Solidity (the client) and another one in Go (the server).
+
+The Solidity binding, also known as `SuaveLib` is a Solidity library that Suave apps can use to call the precompiles. The Golang counterpart runs in the EVM and handles the Solidity calls to the SUAVE precompiles. It generates an skeleton interface that the developer needs to implement with the logic of the precompiles.
+
+The bindings abstract from the developer any encoding/decoding and error management and provide an standard format for both runtimes to communicate with each other.
+
+You can find the full `yaml` specification [here](../gen/suave_spec.yaml).
+
+## Specification
+
+This is the specification of the yaml file.
+
+```yaml
+types:
+  - name: BidId
+    type: bytes16
+structs:
+  - name: Bid
+    fields:
+      - name: id
+        type: BidId
+      - name: decryptionCondition
+        type: uint64
+...
+functions:
+  - name: confidentialInputs
+    address: "0x0000000000000000000000000000000042010001"
+    output:
+      plain: true
+      fields:
+        - name: output1
+          type: bytes
+  - name: newBid
+    address: "0x0000000000000000000000000000000042030000"
+    input:
+      - name: decryptionCondition
+        type: uint64
+      - name: allowedPeekers
+        type: address[]
+      - name: bidType
+        type: string
+    output:
+      fields:
+        - name: bid
+          type: Bid
+```
+
+- types: List of user-defined value types:
```

Comment:
I adapted this section quite extensively when putting it up as a guide on the docs site: https://suave-docs.vercel.app/reference/precompiles/write-your-own